### PR TITLE
feat: Add fault proof support to opstack utils

### PR DESCRIPTION
**Pull request by roninjin10**
Adds fault proof support to viem/opstack in a forwards and backwards compatible way.

---

## PR-Codex overview
The focus of this PR is to add support for Optimism's devnet setup and related contracts. 

### Detailed summary
- Added `.optimism` directory for devnet setup
- Updated `solc` version to `0.8.20`
- Implemented functions to clone Optimism repo and run devnet
- Modified contract addresses and actions related to dispute games and portals

> The following files were skipped due to too many changes: `src/chains/opStack/actions/getPortalVersion.ts`, `src/chains/opStack/actions/getTimeToFinalize.ts`, `src/chains/opStack/actions/getDisputeGame.ts`, `src/chains/opStack/actions/getL2Output.ts`, `src/chains/opStack/actions/getWithdrawalStatus.test.ts`, `src/chains/opStack/actions/getWithdrawalStatus.ts`, `src/chains/opStack/abis.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: scripts/opDevnetTests.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  echo "cloning ${OPTIMISM_DIR} successful"
+fi
+
+echo "starting devnet..."
+cd ${OPTIMISM_DIR}
+pnpm && pnpm build && make devnet-up || { echo 'unable to start devnet' ; exit 1; }
+cd ..
+
+echo "Running tests against optimism devnet..."
+OP_DEVNET=true CI=true vitest -c ./test/vitest.op.config.ts --coverage --retry=3 --bail=1 --pool=forks
```
Comment (roninjin10) [inline scripts/opDevnetTests.sh:19]: This script is spinning up an op devnet to run fault proof tests against
We can remove this in favor of forking sepolia on wednesday

File: src/chains/opStack/actions/buildProveWithdrawal.ts [hunk 2]
```diff
@@ -117,6 +121,28 @@ export async function buildProveWithdrawal<
   BuildProveWithdrawalReturnType<chain, account, chainOverride, accountOverride>
 > {
   const { account, chain = client.chain, output, withdrawal } = args
+
+  const portalAddress = (() => {
+    if (args.portalAddress) return args.portalAddress
+    if (args.chain)
+      return args.targetChain!.contracts.portal[args.chain.id].address
+    return Object.values(args.targetChain!.contracts.portal)[0].address
+  })()
```
Comment (roninjin10) [inline src/chains/opStack/actions/buildProveWithdrawal.ts:130]: The above anvil barf is real and also indicates I accidentally made a breaking change. Will give this a look in a sec

File: src/chains/opStack/actions/buildProveWithdrawal.ts [hunk 3]
```diff
@@ -132,13 +158,41 @@ export async function buildProveWithdrawal<
...
+    portalAddress,
+    strategy: 'random',
+    l2BlockNumber: args.output.l2BlockNumber,
+    disputeGameFactoryAddress,
+  })
+
   return {
     account,
-    l2OutputIndex: output.outputIndex,
+    l2OutputIndex: game.index,
```
Comment (roninjin10) [inline src/chains/opStack/actions/buildProveWithdrawal.ts:191]: Yup should be handling null case here

File: src/chains/opStack/actions/buildProveWithdrawal.ts [hunk 4]
```diff
@@ -132,13 +158,41 @@ export async function buildProveWithdrawal<
...
     outputRootProof: {
       latestBlockhash: block.hash,
       messagePasserStorageRoot: proof.storageHash,
-      stateRoot: block.stateRoot,
+      stateRoot: game.rootClaim,
```
Comment (roninjin10) [inline src/chains/opStack/actions/buildProveWithdrawal.ts:195]: same handle nulll case

File: src/chains/opStack/actions/getDisputeGame.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
+import { decodeAbiParameters } from '~viem/index.js'
```
Comment (roninjin10) [inline src/chains/opStack/actions/getDisputeGame.ts:1]: fixme import

File: src/chains/opStack/actions/getL2Output.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import { decodeAbiParameters } from '~viem/index.js'
```
Comment (roninjin10) [inline src/chains/opStack/actions/getL2Output.ts:1]: fixme import

File: src/chains/opStack/actions/getTimeToFinalize.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
+import { readContract } from '~viem/actions/index.js'
```
Comment (roninjin10) [inline src/chains/opStack/actions/getTimeToFinalize.ts:1]: fixme import

File: src/chains/opStack/actions/buildProveWithdrawal.ts (hunk 2 above)
Comment (roninjin10) [reply src/chains/opStack/actions/buildProveWithdrawal.ts:130]: This is still a bug I think but will debug the broken test in a bit

**Comment by jxom** [comment]
Closing in favor of #1983

**Pull request closed without merging**